### PR TITLE
[python] Use the same timestamp across read and write operations

### PR DESCRIPTION
**Pull request by thetorpedodog**
This eliminates the distinction between the timestamps used to read and write TileDB data. This ensures that:

- All data is written with the same timestamp.
- When one context is used to write and then read the same data, the same timestamp is used for both operations.
- A consistent view of a TileDB object is used for metadata when opening a handle for writing.

---

Given that all our tests pass, and this eliminates the "auxiliary read at time now" hack, I think this might be the right thing to do for this case.

### CODE REVIEW

File: apis/python/src/tiledbsoma/_tdb_handles.py
Comment (mlin) [inline apis/python/src/tiledbsoma/_tdb_handles.py:223]: I think this is kind of a thorny question, if for whatever reason we've opened the object for writing at a *past* timestamp (preceding other, existing, later-timestamped writes), what view of the contents/metadata should we start with? But that's a weird enough thing to be doing in the first place that I'm not too worried about it. Simplifying the code is good.

**Review by mlin — APPROVED** [review]
Subjective opinions written in slack. No objections if this is the direction we want to go. Note this is a late, breaking change to 1.0rc0 for @atolopko-czi @bkmartinjr

**Comment by thetorpedodog** [comment]
preserving some of the stuff I have said in chat for posterity:

The model I have in my head, though, is that when interacting with a dataset, you are interacting with TileDB at *one specific timestamp*. If you want to read at a different timestamp, that belongs in a different context. (This would be even more salient with the proposed reopen API.)

Moreover, based on the following events:

- 10:00: Create collection `a`. This sets the write timestamp of all changes we do to `a`.
- 10:02: `a.add_new_dataframe('b', ...)`. Write lots of data to `b`.
- 10:04: `a.add_new_dataframe('c', ...)`. Write lots more data to `c`.
- 10:06: Close everything.

With the existing model, the timestamp of a write is determined by when the open occurs. That means that the write to `a`, which includes adding `b` and `c` as sub-elements, occurs at 10:00. However, the writes to `b` and `c` only occur at 10:02 and :04 (respectively), meaning that, if after all writes complete:

- If you open `a` as of 10:01, `b` and `c` will both appear as members of the group, but *will not contain any data* (in fact, the metadata will not be set, which I am now thinking “write metadata with the timestamp of object creation” is probably the right thing to do).
- If you open `a` as of 10:03, only `b` will contain data.
- Only if you open `a` as of 10:04 or later, will the full object be there.

With this new model, all of `a`, `b`, and `c` will be created and written as of the same timestamp (10:00, by default, when the context was created).

> I think this is kind of a thorny question, if for whatever reason we've opened the object for writing at a past timestamp (preceding other, existing, later-timestamped writes), what view of the contents/metadata should we start with?

My strong opinion is that it should always be as of the only specified timestamp, so that you get a temporally coherent view no matter what. If the user wants to act based on the “current” metadata, they can open the array/group/whatever as of Now (or The Infinite Future, as the case may be).

I’m going to leave this open until sometime tomorrow so that more discussion can happen if need be.

**Comment by mlin** [comment]
Then also for the record, my opinion is that setting the same timestamp across a group of writes is currently not advisable (especially as a default behavior) due to the randomization behavior exhibited in #919. We *think* we know how to avoid that (TBD -- it [may not yet be fixed](https://github.com/chanzuckerberg/cell-census/pull/210)), but it's a deep and wide pitfall, which we ourselves fell into without delay. Again, I defer to you as long as this has been considered and deemed an acceptable risk (hopefully a temporary one).

**Comment by thetorpedodog** [comment]
What I don’t entirely understand is what circumstances are leading to randomized writes—are you opening, writing, and closing the same SOMA object multiple times over the course of a single execution? Because if you’re doing multiple writes on the same `open`, the behavior here will be unchanged on a per–SOMA object basis, since all write calls on the same `open` are recorded at the same timestamp anyway.

The other problem I was aiming to solve here is that there is an asymmetry between the way read timestamps and write timestamps are handled. For instance, with the previous behavior:

1. T+0:00: Create a new context with default options.
2. T+0:01: Use the context to write a bunch of data. Close.
3. T+0:02: Use that same context to *read* the data you just wrote.

When you read the data, it will not actually see the data you just wrote, because the read timestamp stored in the context is T+0:00, before the (automatic) timestamp used to write the data at T+0:01.

My feeling is that if we’re writing at the current timestamp, we should also be *reading* at the current timestamp, so the behavior is consistent on both sides. This can be achieved by allowing the user to opt out of timestamp consistency by passing `None` as a timestamp. (Alternately, we could make it opt-in, where `None` is the default but a user can pass in the current timestamp at context creation.)

**Comment by thetorpedodog** [comment]
Issuing a correction on my previous statements regarding write timing: The timestamps of writes are recorded as of the **close** call, not as of the open. This means that some of the worse symptoms described above do not occur, but the asymmetry between read and write would remain.

**Comment by mlin** [comment]
My caution is an acknowledgment that I don't think I fully understand all the circumstances when it may be hazardous to set the same timestamp across my logical group of writes -- it's already surprised me a couple of times. Since we're demonstrably in the vicinity of the worst possible kind of bug -- nondeterministic, silent, and undetected data corruption -- my instinct is to just avoid the whole area, if possible, rather than attempt to navigate it safely.

On the other side of the scale, I understand we want to share the write timestamps mainly so that there will exist no timestamp when a reader would see some but not all of our writes. That's a virtuous property, which however I don't think necessary for SOMA & Cell Census. It should suffice that we can publish a known-good timestamp when the reader will see all of our writes (and nothing written after). I think we can simply let the timestamps default to "now" as we actually apply each write, per TileDB default behavior, and report when we finished.

Therefore I wouldn't set any shared write timestamp by default given our experience with the current version of TileDB. That implies I think the write timestamp is distinct from the read timestamp, where a shared default seems reasonable for some snapshot consistency (but always better to set a known-good one explicitly). 

This isn't to say the current implementation doesn't have its own sharp edges that could be improved, which you've pointed out, but I feel your full model for read/write temporal coherence across a group of objects would be more safely adopted by TileDB-SOMA at a future date, when it's fully supported by TileDB itself, without even the slightest proximity to a data corruption hazard area.

Finally, to belabor it once again, this is all my personal opinion; I'm not suggesting I'm right, and I'm fine with agreeing to disagree on any of it, with you choosing what to do next.

**Comment by thetorpedodog** [comment]
Getting the opt-out tested ended up taking a little more time than I anticipated (mostly, the part where I was learning more about how timestamps work that led to my last comment), so I wasn’t able to get it out yesterday.

*For now* I would still prefer to make consistency opt-out, but we still have the option to make it opt-in if we need to before release.

**Review by mlin — APPROVED** [review]
With aforementioned discomfort -- agreeing to disagree

**Comment by thetorpedodog** [comment]
After some discussion with Bruce, I’m going to submit this with timestamp consistency being the default (so that by doing nothing you get consistency). I would like to discuss at the big meeting tomorrow whether that *should* be the default, and if we want to go to “no timestamp by default” it will be a one-line change, just so that the rest of the implementation isn’t blocked on the default decision.

**Pull request merged**
